wgpu-core/src/device/mod.rs

```diff
@@ -182,15 +182,37 @@ impl UserClosures {
             closure();
         }
         for invocation in self.device_lost_invocations {
-            (invocation.closure)(invocation.reason, invocation.message);
+            invocation
+                .closure
+                .call(invocation.reason, invocation.message);
         }
     }
 }
 
 #[cfg(send_sync)]
-pub type BoxedDeviceLostClosure = Box<dyn FnOnce(DeviceLostReason, String) + Send + 'static>;
+pub type BoxedDeviceLostClosure = Box<dyn DeviceLostClosure + Send + 'static>;
 #[cfg(not(send_sync))]
-pub type BoxedDeviceLostClosure = Box<dyn FnOnce(DeviceLostReason, String) + 'static>;
+pub type BoxedDeviceLostClosure = Box<dyn DeviceLostClosure + 'static>;
+
+/// This trait is only needed because implementing [`FnOnce`] on user types is not possible right now
+/// due to `fn_traits` being unstable, see <https://github.com/rust-lang/rust/issues/29625>.
+///
+/// This means that an extra indirection is needed in cases where a `Box<dyn FnOnce>`
+/// would have sufficed. Wrap a `Box<dyn FnOnce>` in a [`Box`] to get a `Box<dyn DeviceLostClosure>`.
+pub trait DeviceLostClosure {
+    fn call(self: Box<Self>, reason: DeviceLostReason, message: String);
+}
```

Comment:
question: I'm actually sorta confused; why is the `DeviceLostClosure` trait interesting, as opposed to "just" capturing something in a `FnOnce` with a custom `Drop` implementation?

Reply:
The issue is that the `FnOnce` might never be called.

Reply:
Actually, would that still work fine? I guess the `Drop` impl would still have to run even if the closure is dropped.

Reply:
I will try it in rust playground just to make sure, but this sounds hopeful, thanks for bringing it up!

Reply:
It should still work fine; `Drop` of captured variables should work just like if we use a custom trait!